### ISSUE TITLE
Same seed but different results

### ISSUE THREAD

**Issue by HenriqueAJNB** [microsoft/LightGBM#4722]
## Description
LightGBM brings different results when using same same seed

## Reproducible example
Can't share much details because it's not an open project.

This is how I'm training model in different platforms, with same dataset
```python
params = {
    "boosting_type": "goss",
    "objective": "mean_absolute_error",
    "metric": "mape",
    "subsample_freq": 0,
    "learning_rate": 0.1,
    "min_data_in_leaf": 1,
    "feature_fraction": 1,
    "boost_from_average": True,
    "verbose": -1,
    "max_bin": 100,
    "num_leaves": 100072,
    "tree_learner": "voting",
    "seed": 42,
}

model = lightgbm.train(
    params,
    lightgbm.Dataset(X_train, y_train, categorical_feature=categorical_feature),
    num_boost_round=100,
)
```
The dataset has a shape of `(6613, 58)` and only 3 categorical features, transformed into categorical using 
```python
df[categorical_feature] = df[categorical_feature].astype("category")
```

## Environment info

LightGBM version or commit hash: 3.2.1

Command(s) you used to install LightGBM

```shell
pip install lightgbm==3.2.1
```

Here I'll specify most that I can about platform 1 and 2 that are running the same training code above:

**Platform 1**
- **OS:** Microsoft Windows 10 Enterprise x64 architecture
- **Python version**: 3.9.7
- **CPU:**  Intel64 Family 6 Model 140
- **Pandas version:** 1.3.3
- **Numpy version:** 1.21.2

**Platform 2**
- **OS:** Microsoft Windows Server 2012  x64 architecture
- **Python version**: 3.7.9
- **CPU:**  Intel64 Family 6 Model 79
- **Pandas version:** 1.3.3
- **Numpy version:** 1.21.2

## Additional Comments
Both models are running on VSCode 1.61 IDE within an specific virtual environment for each platform

## Results

Here are the `.describe()` of column containing the predictions for each of the platforms.

```
count    6613.000000
mean        3.730982
std         1.229532
min        -0.001069
25%         2.932345
50%         3.841217
75%         4.611050
max         6.901085
Name: PRED_PLAT1, dtype: float64

count    6613.000000
mean        3.744200
std         1.214541
min         0.000026
25%         2.919000
50%         3.899890
75%         4.612596
max         6.851869
Name: PRED_PLAT2, dtype: float64
```

Notice that they are quite close, but not identical, as I expected to be using `seed=42` in hyperparameters.

Next try will be using `deterministic=True`, but hopeless.
